doc/pbh/pbh-design.md

```diff
@@ -0,0 +1,758 @@
+        SwitchOrch *switchOrch,
+        PortsOrch *portOrch
+    );
+    ~PbhOrch();
+
+    using Orch::doTask;  // Allow access to the basic doTask
+
+private:
+    void doPbhTableTask(Consumer &consumer);
+    void doPbhRuleTask(Consumer &consumer);
+    void doPbhHashTask(Consumer &consumer);
+    void doTask(Consumer &consumer);
+
+    SwitchOrch *m_switchOrch;
+    PortsOrch *m_portOrch;
+};
+```
+
+### 2.3.3 ACL orch
+
+This orchestrator provides API for ACL table/rule configuration.  
+It is already exists in SONiC.
+
+ACL orchestrator will be extended to support PBH table/rule concept.  
+PBH table will use a dedicated set of keys to allow match of NVGRE and VxLAN packets.
+
+**Skeleton code:**
+```cpp
+bool AclTable::create()
+{
+    ...
+
+    if (type == ACL_TABLE_PBH)
+    {
+        attr.id = SAI_ACL_TABLE_ATTR_FIELD_GRE_KEY;
+        attr.value.booldata = true;
+        table_attrs.push_back(attr);
+
+        attr.id = SAI_ACL_TABLE_ATTR_FIELD_IP_PROTOCOL;
+        attr.value.booldata = true;
+        table_attrs.push_back(attr);
+
+        attr.id = SAI_ACL_TABLE_ATTR_FIELD_L4_DST_PORT;
+        attr.value.booldata = true;
+        table_attrs.push_back(attr);
+
+        attr.id = SAI_ACL_TABLE_ATTR_FIELD_INNER_ETHER_TYPE;
+        attr.value.booldata = true;
+        table_attrs.push_back(attr);
+
+        attr.id = SAI_ACL_TABLE_ATTR_ACL_STAGE;
+        attr.value.s32 = SAI_ACL_STAGE_INGRESS;
+        table_attrs.push_back(attr);
+
+        sai_status_t status = sai_acl_api->create_acl_table(&m_oid, gSwitchId, (uint32_t)table_attrs.size(), table_attrs.data());
+
+        if (status == SAI_STATUS_SUCCESS)
+        {
+            gCrmOrch->incCrmAclUsedCounter(CrmResourceType::CRM_ACL_TABLE, acl_stage, SAI_ACL_BIND_POINT_TYPE_PORT);
+            gCrmOrch->incCrmAclUsedCounter(CrmResourceType::CRM_ACL_TABLE, acl_stage, SAI_ACL_BIND_POINT_TYPE_LAG);
+        }
+
+        return status == SAI_STATUS_SUCCESS;
+    }
+
+    ...
+}
+
+class AclRulePbh: public AclRule
+{
+public:
+    AclRulePbh(AclOrch *m_pAclOrch, string rule, string table, acl_table_type_t type, bool createCounter = false);
+
+    bool validateAddAction(string attr_name, string attr_value);
+    bool validateAddMatch(string attr_name, string attr_value);
+    bool validate();
+    void update(SubjectType, void *);
+};
+```
+
+## 2.4 DB schema
+
+### 2.4.1 Config DB
+
+#### 2.4.1.1 PBH table
+```abnf
+; defines schema for PBH table configuration attributes
+key = PBH_TABLE|table_name ; table name. Must be unique
+
+; field     = value
+port_list   = port-list ; ports to which this table is applied. Can be empty
+lag_list    = lag-list  ; portchannels to which this table is applied. Can be empty
+description = *255VCHAR ; table description. Can be empty
+
+; value annotations
+port-name = 1*64VCHAR                         ; name of the port
+port-list = port-name [ 1*( "," port-name ) ] ; list of the ports. Valid values range is platform dependent
```

Comment:
We dont need port-name when port-list itself can have a single port-name..correct?